server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/cache/CachingRepository.java

```diff
@@ -113,10 +106,40 @@ public CompletableFuture<Map<String, Entry<?>>> find(Revision revision, String p
         requireNonNull(options, "options");
 
         final Revision normalizedRevision = normalizeNow(revision);
-        return cache.get(new CacheableFindCall(repo, normalizedRevision, pathPattern, options))
-                    .handleAsync((unused, cause) -> {
+
+        Map<FindOption<?>, ?> cacheableOptions = options;
+        final Integer maxEntries = (Integer) options.get(FindOption.MAX_ENTRIES);
+        if (maxEntries != null) {
+            final ImmutableMap.Builder<FindOption<?>, Object> newOptions = ImmutableMap.builder();
+            options.forEach((key, value) -> {
+                if (key != FindOption.MAX_ENTRIES) {
+                    newOptions.put(key, value);
+                }
+            });
+            cacheableOptions = newOptions.build();
+        }
+
+        return cache.get(new CacheableFindCall(repo, normalizedRevision, ALL_PATH, cacheableOptions))
+                    .handleAsync((all, cause) -> {
                         throwUnsafelyIfNonNull(cause);
-                        return unused;
+
+                        if (all.isEmpty()) {
+                            return all;
+                        }
+
+                        Stream<Map.Entry<String, Entry<?>>> stream = all.entrySet().stream();
+                        if (!pathPattern.equals(ALL_PATH)) {
+                            final PathPatternFilter filter = PathPatternFilter.of(pathPattern);
+                            stream = stream.filter(entry -> filter.matches(entry.getKey()));
+                        }
+                        if (maxEntries != null) {
+                            stream = stream.limit(maxEntries);
+                        }
+
+                        // Use LinkedHashMap to 1) keep the order and 2) allow callers to mutate it.
+                        return stream.collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue,
+                                                               (oldV, newV) -> oldV,
```

Comment:
Question) Is this just a precaution? or is it actually possible that multiple values (files) exist for a single key (path)?

Reply:
`(oldV, newV) -> oldV` was just added because of the signature of `Collectors.toMap()`.
https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collectors.html#toMap-java.util.function.Function-java.util.function.Function-java.util.function.BinaryOperator-java.util.function.Supplier-

To use `LinkedHashMap`, `mergeFunction` is necessary.

Reply:
> allow callers to mutate it.

Is this necessary? I'm asking because we used to return the map as is.
Also what do you think of returning the map as-is when `maxEntries` is `null` so that we don't do the additional `stream.collect` call.

Reply:
> Is this necessary?

Our current implementation mutates the returned entries.
https://github.com/line/centraldogma/blob/3a261ba4ec9685b27d5287b239503df4bce767e5/server-mirror-git/src/main/java/com/linecorp/centraldogma/server/internal/mirror/AbstractGitMirror.java#L356-L366

Reply:
>  what do you think of returning the map as-is when maxEntries is null 

Did you mean `pathPattern == ALL_PATH && maxEntries == null`? As the caller mutates it, creating a new collection would be safer than returning the cached one.

Anyway, I believe using immutable data is a good approach. We need to fix the callers to create a new collection to mutate. I will handle it in the following PR.

Reply:
> Anyway, I believe using immutable data is a good approach. We need to fix the callers to create a new collection to mutate. I will handle it in the following PR.

I agree. I didn't expect that the call would modify the map. (I think I implemented it in that way 😓 )